File: text_editor.py
import tkinter as tk
from tkinter import filedialog, messagebox, font
import os

class TextEditor:
    def __init__(self, root):
        self.root = root
        self.root.title("Notepad")
        self.root.geometry("600x400")

        self.text_area = tk.Text(self.root, undo=True)
        self.text_area.pack(fill=tk.BOTH, expand=1)

        self.current_file = None

        self.create_menu()
        
    def create_menu(self):
        menu_bar = tk.Menu(self.root)
        self.root.config(menu=menu_bar)

        file_menu = tk.Menu(menu_bar, tearoff=0)
        menu_bar.add_cascade(label="File", menu=file_menu)
        file_menu.add_command(label="New", command=self.new_file)
        file_menu.add_command(label="Open", command=self.open_file)
        file_menu.add_command(label="Save", command=self.save_file)
        file_menu.add_command(label="Save As", command=self.save_as_file)
        file_menu.add_separator()
        file_menu.add_command(label="Exit", command=self.exit_app)

        edit_menu = tk.Menu(menu_bar, tearoff=0)
        menu_bar.add_cascade(label="Edit", menu=edit_menu)
        edit_menu.add_command(label="Undo", command=self.text_area.edit_undo)
        edit_menu.add_command(label="Redo", command=self.text_area.edit_redo)
        edit_menu.add_separator()
        edit_menu.add_command(label="Cut", command=self.cut_text)
        edit_menu.add_command(label="Copy", command=self.copy_text)
        edit_menu.add_command(label="Paste", command=self.paste_text)

        format_menu = tk.Menu(menu_bar, tearoff=0)
        menu_bar.add_cascade(label="Format", menu=format_menu)
        format_menu.add_command(label="Font", command=self.select_font)
        format_menu.add_command(label="Font Size", command=self.select_font_size)
        format_menu.add_command(label="Bold", command=self.toggle_bold)
        format_menu.add_command(label="Italic", command=self.toggle_italic)

    def new_file(self):
        self.text_area.delete(1.0, tk.END)
        self.current_file = None

    def open_file(self):
        file_path = filedialog.askopenfilename(defaultextension=".txt",
                                               filetypes=[("Text files", "*.txt"), ("All files", "*.*")])
        if file_path:
            with open(file_path, "r") as file:
                self.text_area.delete(1.0, tk.END)
                self.text_area.insert(tk.END, file.read())
            self.current_file = file_path

    def save_file(self):
        if self.current_file:
            try:
                with open(self.current_file, "w") as file:
                    file.write(self.text_area.get(1.0, tk.END))
            except Exception as e:
                messagebox.showerror("Save Error", str(e))
        else:
            self.save_as_file()

    def save_as_file(self):
        file_path = filedialog.asksaveasfilename(defaultextension=".txt",
                                                 filetypes=[("Text files", "*.txt"), ("All files", "*.*")])
        if file_path:
            try:
                with open(file_path, "w") as file:
                    file.write(self.text_area.get(1.0, tk.END))
                self.current_file = file_path
            except Exception as e:
                messagebox.showerror("Save As Error", str(e))

    def exit_app(self):
        self.root.quit()

    def cut_text(self):
        self.text_area.event_generate("<<Cut>>")

    def copy_text(self):
        self.text_area.event_generate("<<Copy>>")

    def paste_text(self):
        self.text_area.event_generate("<<Paste>>")

    def select_font(self):
        fonts = list(font.families())
        selected_font = tk.simpledialog.askstring("Select Font", "Font:", initialvalue=self.text_area.cget("font"))
        if selected_font in fonts:
            self.text_area.config(font=(selected_font, self.text_area.cget("font")[1]))

    def select_font_size(self):
        size = tk.simpledialog.askinteger("Font Size", "Enter Font Size:", initialvalue=self.text_area.cget("font")[1])
        if size:
            self.text_area.config(font=(self.text_area.cget("font")[0], size))

    def toggle_bold(self):
        current_font = font.Font(font=self.text_area.cget("font"))
        if current_font.actual()["weight"] == "bold":
            new_font = current_font.copy()
            new_font.config(weight="normal")
        else:
            new_font = current_font.copy()
            new_font.config(weight="bold")
        self.text_area.config(font=new_font)

    def toggle_italic(self):
        current_font = font.Font(font=self.text_area.cget("font"))
        if current_font.actual()["slant"] == "italic":
            new_font = current_font.copy()
            new_font.config(slant="roman")
        else:
            new_font = current_font.copy()
            new_font.config(slant="italic")
        self.text_area.config(font=new_font)

if __name__ == "__main__":
    root = tk.Tk()
    app = TextEditor(root)
    root.mainloop()

File: TextEditor.java
import javax.swing.*;
import javax.swing.undo.*;
import java.awt.*;
import java.awt.event.*;
import java.io.*;

public class TextEditor extends JFrame implements ActionListener {
    private JTextArea textArea;
    private JFileChooser fileChooser;
    private String currentFile;
    private Font currentFont;
    private UndoManager undoManager;

    public TextEditor() {
        setTitle("Notepad");
        setSize(600, 400);

        textArea = new JTextArea();
        JScrollPane scrollPane = new JScrollPane(textArea);
        add(scrollPane);

        JMenuBar menuBar = new JMenuBar();
        JMenu fileMenu = new JMenu("File");
        JMenuItem newMenuItem = new JMenuItem("New");
        JMenuItem openMenuItem = new JMenuItem("Open");
        JMenuItem saveMenuItem = new JMenuItem("Save");
        JMenuItem saveAsMenuItem = new JMenuItem("Save As");
        JMenuItem exitMenuItem = new JMenuItem("Exit");

        newMenuItem.addActionListener(this);
        openMenuItem.addActionListener(this);
        saveMenuItem.addActionListener(this);
        saveAsMenuItem.addActionListener(this);
        exitMenuItem.addActionListener(this);

        fileMenu.add(newMenuItem);
        fileMenu.add(openMenuItem);
        fileMenu.add(saveMenuItem);
        fileMenu.add(saveAsMenuItem);
        fileMenu.add(exitMenuItem);
        menuBar.add(fileMenu);

        JMenu formatMenu = new JMenu("Format");
        JMenuItem fontMenuItem = new JMenuItem("Font");
        JMenuItem fontSizeMenuItem = new JMenuItem("Font Size");
        JMenuItem boldMenuItem = new JMenuItem("Bold");
        JMenuItem italicMenuItem = new JMenuItem("Italic");

        fontMenuItem.addActionListener(this);
        fontSizeMenuItem.addActionListener(this);
        boldMenuItem.addActionListener(this);
        italicMenuItem.addActionListener(this);

        formatMenu.add(fontMenuItem);
        formatMenu.add(fontSizeMenuItem);
        formatMenu.add(boldMenuItem);
        formatMenu.add(italicMenuItem);
        menuBar.add(formatMenu);

        JMenu editMenu = new JMenu("Edit");
        JMenuItem undoMenuItem = new JMenuItem("Undo");
        JMenuItem redoMenuItem = new JMenuItem("Redo");
        JMenuItem cutMenuItem = new JMenuItem("Cut");
        JMenuItem copyMenuItem = new JMenuItem("Copy");
        JMenuItem pasteMenuItem = new JMenuItem("Paste");

        undoMenuItem.addActionListener(this);
        redoMenuItem.addActionListener(this);
        cutMenuItem.addActionListener(this);
        copyMenuItem.addActionListener(this);
        pasteMenuItem.addActionListener(this);

        editMenu.add(undoMenuItem);
        editMenu.add(redoMenuItem);
        editMenu.add(cutMenuItem);
        editMenu.add(copyMenuItem);
        editMenu.add(pasteMenuItem);
        menuBar.add(editMenu);

        setJMenuBar(menuBar);
        fileChooser = new JFileChooser();
        currentFont = new Font("Arial", Font.PLAIN, 12);
        undoManager = new UndoManager();
        textArea.getDocument().addUndoableEditListener(undoManager);

        setVisible(true);
    }

    public void actionPerformed(ActionEvent e) {
        String command = e.getActionCommand();

        if (command.equals("New")) {
            textArea.setText("");
            currentFile = null;
        } else if (command.equals("Open")) {
            int returnVal = fileChooser.showOpenDialog(this);
            if (returnVal == JFileChooser.APPROVE_OPTION) // By selecting OK
            {
                File file = fileChooser.getSelectedFile();
                try {
                    FileReader reader = new FileReader(file);
                    textArea.read(reader, null); // Reading file line by line
                    reader.close();
                    currentFile = file.getAbsolutePath(); // Setting path
                } catch (IOException ex) {
                    ex.printStackTrace();
                }
            }
        } else if (command.equals("Save")) {
            if (currentFile != null) {
                try {
                    FileWriter writer = new FileWriter(currentFile);
                    textArea.write(writer);
                    writer.close();
                } catch (IOException ex) {
                    ex.printStackTrace();
                }
            } else {
                saveAs();
            }
        } else if (command.equals("Save As")) {
            saveAs();
        } else if (command.equals("Exit")) {
            System.exit(0);
        } else if (command.equals("Font")) {
            selectFont();
        } else if (command.equals("Font Size")) {
            selectFontSize();
        } else if (command.equals("Bold")) {
            toggleBold();
        } else if (command.equals("Italic")) {
            toggleItalic();
        } else if (command.equals("Undo")) {
            if (undoManager.canUndo()) {
                undoManager.undo();
            }
        } else if (command.equals("Redo")) {
            if (undoManager.canRedo()) {
                undoManager.redo();
            }
        } else if (command.equals("Cut")) {
            textArea.cut();
        } else if (command.equals("Copy")) {
            textArea.copy();
        } else if (command.equals("Paste")) {
            textArea.paste();
        }
    }

    private void saveAs() {
        int returnVal = fileChooser.showSaveDialog(this);
        if (returnVal == JFileChooser.APPROVE_OPTION) {
            File file = fileChooser.getSelectedFile();
            try {
                FileWriter writer = new FileWriter(file);
                textArea.write(writer);
                writer.close();
                currentFile = file.getAbsolutePath();
            } catch (IOException ex) {
                ex.printStackTrace();
            }
        }
    }

    private void selectFont() {
        String[] fonts = GraphicsEnvironment.getLocalGraphicsEnvironment().getAvailableFontFamilyNames();
        String selectedFontName = (String) JOptionPane.showInputDialog(
                this,
                "Select Font",
                "Font",
                JOptionPane.PLAIN_MESSAGE,
                null,
                fonts,
                currentFont.getFontName());
        if (selectedFontName != null) {
            currentFont = new Font(selectedFontName, currentFont.getStyle(), currentFont.getSize());
            textArea.setFont(currentFont);
        }
    }

    private void selectFontSize() {
        String sizeStr = JOptionPane.showInputDialog(this, "Enter Font Size:");
        if (sizeStr != null && !sizeStr.isEmpty()) {
            int size = Integer.parseInt(sizeStr);
            if (size > 0) {
                currentFont = currentFont.deriveFont((float) size);
                textArea.setFont(currentFont);
            }
        }
    }

    private void toggleBold() {
        int style = currentFont.getStyle();
        if ((style & Font.BOLD) == Font.BOLD) {
            style &= ~Font.BOLD; // Clear the bold
        } else {
            style |= Font.BOLD; // Set the bold
        }
        currentFont = currentFont.deriveFont(style);
        textArea.setFont(currentFont);
    }

    private void toggleItalic() {
        int style = currentFont.getStyle();
        if ((style & Font.ITALIC) == Font.ITALIC) {
            style &= ~Font.ITALIC; // Clear the italic
        } else {
            style |= Font.ITALIC; // Set the italic
        }
        currentFont = currentFont.deriveFont(style);
        textArea.setFont(currentFont);
    }

    public static void main(String[] args) {
        SwingUtilities.invokeLater(() -> new TextEditor());
    }
}
